replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,259 @@
+        if (supported(task)) {
+          this.handle(task);
+        }
+        this.processing.set(false);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Set the running flag to false and wait for the thread to die. If an item is currently in the
+   * process of transferring, it will finish.
+   */
+  public void cancel() {
+    this.running = false;
+    try {
+      Thread.currentThread().join();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  /**
+   * @return {@code true} if this worker is transferring an item, otherwise {@code false}
+   */
+  public boolean isProcessing() {
+    return processing.get();
+  }
+
+  private boolean supported(Task task) throws InterruptedException {
+    if (!OperationType.UNKNOWN.equals(task.getOperation())) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          String.format("Operation task %s is not supported by this worker", task));
+      return false;
+    }
+    return true;
+  }
+
+  private void handle(Task task) throws InterruptedException {
+    boolean proceed = checkAdapterAvailabilities(task);
+    if (!proceed) {
+      return;
+    }
+
+    List<DDFMetadataInfo> infos = task.metadatas()
+        .filter(t -> t instanceof DDFMetadataInfo)
+        .map(DDFMetadataInfo.class::cast)
+        .collect(Collectors.toList());
+
+    if (infos.size() > 1) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          "more than one metadata is currently not supported by this worker");
+      return;
+    }
+
+    if(infos.isEmpty()) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          "unable to handle any of the metadatas received in this task");
+      return;
+    }
+
+    final DDFMetadataInfo info = infos.get(0);
+    final Metadata metadata = new MetadataImpl(((Map) info.getData()).get("raw-metadata"),
+        info.getDataClass(), task.getId(),
+        Date.from(info.getLastModified()));
+
+    final OperationType type = task.getOperation();
+    switch (type) {
+      case HARVEST:
+      case REPLICATE_FROM:
+        doCreate(task, metadata, destination, local);
+        break;
+      case REPLICATE_TO:
+        doCreate(task, metadata, local, destination);
+        break;
+      case DELETE:
+        // always delete on non-local site
+        doDelete(task, metadata);
+        break;
+      default:
+        // should never hit this since deserialization will return UNKNOWN
+        task.fail(ErrorCode.UNSUPPORTED_ERROR,
+            String.format("Operation task %s is not supported by this worker", task));
+    }
+  }
+
+  private void doCreate(final Task task, final Metadata metadata, final NodeAdapter source,
+      final NodeAdapter destination)
+      throws InterruptedException {
+    final String metadataId = task.getId();
+    final String operation = task.getOperation().name();
+    if (source.exists(metadata)) {
```

Comment:
❓ Are we going to always check first? Does that now mean that for all normal cases we shall end up with 2 requests to the source, one to check and one to get the data? Can we not take the approach where we try to get it and it either tells us it doesn't have it (e.g. 404-not found) or we get the data?

Reply:
Kept the exist and left a todo with a link to https://github.com/connexta/replication/issues/235